grants_tagger_light/retagging/retagging.py

```diff
@@ -0,0 +1,319 @@
+    threshold: float = 0.8,
+    train_examples: int = 100,
+    supervised: bool = True,
+    years: list = None,
+):
+    # We only have 1 file, so no sharding is available https://huggingface.co/docs/datasets/loading#multiprocessing
+    logging.info("Loading the MeSH jsonl...")
+    dset = load_dataset("json", data_files=data_path, num_proc=1)
+    if "train" in dset:
+        dset = dset["train"]
+
+    if years is not None:
+        logger.info(f"Removing all years which are not in {years}")
+        dset = dset.filter(
+            lambda x: any(np.isin(years, [str(x["year"])])), num_proc=num_proc
+        )
+
+    if tags_file_path is not None and os.path.isfile(tags_file_path):
+        with open(tags_file_path, "r") as f:
+            tags = [x.strip() for x in f.readlines()]
+
+    logging.info(f"- Total tags detected: {tags}.")
+    logging.info("- Training classifiers (retaggers)")
+
+    for tag in tags:
+        os.makedirs(os.path.join(save_to_path, tag.replace(" ", "")), exist_ok=True)
+        logging.info(f"- Obtaining positive examples for {tag}...")
+        positive_dset = dset.filter(lambda x: tag in x["meshMajor"], num_proc=num_proc)
+
+        if len(positive_dset["abstractText"]) < train_examples:
+            logging.info(
+                f"Skipping {tag}: low examples "
+                f"({len(positive_dset['abstractText'])} vs "
+                f"expected {train_examples}). "
+                f"Check {save_to_path}.err for more information "
+                f"about skipped tags."
+            )
+            with open(f"{save_to_path}.err", "a") as f:
+                f.write(tag)
+            continue
+        logging.info(f"-- Total positive examples for {tag}: {len(positive_dset)}")
+        logging.info(f"- Obtaining negative examples ('other') for {tag}...")
+        negative_dset = dset.filter(
+            lambda x: tag not in x["meshMajor"], num_proc=num_proc
+        )
+        logging.info(f"-- Total negative examples for {tag}: {len(negative_dset)}")
+
+        curation_file = os.path.join(save_to_path, tag.replace(" ", ""), "curation")
+        if supervised:
+            logging.info(f"- Curating {tag}...")
+            _curate(curation_file, positive_dset, negative_dset, tag, train_examples)
+        else:
+            with open(curation_file, "w") as f:
+                json.dump(
+                    {
+                        tag: {
+                            "positive": [
+                                positive_dset[i] for i in range(train_examples)
+                            ],
+                            "negative": [
+                                negative_dset[i] for i in range(train_examples)
+                            ],
+                        }
+                    },
+                    f,
+                )
+
+    logging.info("- Retagging...")
+
+    models = {}
+    for tag in tags:
+        curation_file = os.path.join(save_to_path, tag.replace(" ", ""), "curation")
```

Comment:
refactor

Reply:
The idea is to do:
1. For each tag, the annotation. We annotate all tags.
2. Again, for each tag, we run classification, but after all annotations have finished.

If I refactor and do both things in 1 loop, it will be:
1. For each tag, annotate and then train. Then jump to the next tag, annotate - train...

It's another approach, but I think it's better to first annotate everything and then train everything.